### PR TITLE
[WIP] Move to pull-streams

### PR DESCRIPTION
**Pull request by dignifiedquire**
Investigate `delay`. Ref https://github.com/pull-stream/pull-ws/issues/13

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,147 +1,53 @@
...
-    return new WebSockets()
-  }
+const createListener = require('./listener')
 
-  this.dial = function (ma, options, callback) {
+module.exports = class WebSockets {
+  dial (ma, options, cb) {
     if (typeof options === 'function') {
-      callback = options
+      cb = options
```
Comment (diasdavid) [inline src/index.js]: Let's continue to use `callback` for top level callback and `cb` for internals (like run-series cb)

Comment (dignifiedquire) [inline src/index.js]: <insert some grumpy comment>

**Comment by diasdavid** [comment]
@whyrusleeping @jbenet Would like your thoughts on a extension to WebSockets to enable halfClosed pipes - https://github.com/pull-stream/pull-ws/issues/13#issuecomment-240722307, it might just be a pattern we need to apply to other transports anyway.

File: src/index.js [hunk 2]
```diff
@@ -1,147 +1,53 @@
...
-
-    conn.getObservedAddrs = (cb) => {
+    const url = `ws://${maOpts.host}:${maOpts.port}`
+    log('dialing %s', url)
+    const socket = connect(url, cb)
+    socket.getObservedAddrs = (cb) => {
       return cb(null, [ma])
     }
 
+    const conn = new Connection(socket)
```
Comment (diasdavid) [inline src/index.js]: Same as https://github.com/libp2p/js-libp2p-tcp/pull/20/files#r75307992

**Comment by dignifiedquire** [comment]
@diasdavid as long as we are using something on top, like a stream muxer that has an end handshake we should be fine. At least as far as I understand

**Comment by diasdavid** [comment]
@dignifiedquire not quite, because spdy also has a closing ceremony (it believes it is on top of TCP), so it will call .end/.close, and not wait to receive the signal to close graciously, causing the same problem.

**Comment by dignifiedquire** [comment]
:cry:

File: src/listener.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    listener._listen(ma.toOptions(), cb)
+  }
+
+  listener.getAddrs = (cb) => {
+    cb(null, [listeningMultiaddr])
+  }
+
+  return listener
```
Comment (diasdavid) [inline src/listener.js]: It is missing to bubble up the events from the listener:
- [ ] event: 'listening'
- [ ] event: 'close'
- [ ] event: 'connection'
- [ ] event: 'error'

See interface-transport.

- [ ] Add tests for the above

**Comment by diasdavid** [comment]
Last things to do:

- [ ] Avoid the monkey patching the socket https://github.com/libp2p/js-libp2p-websockets/pull/18#discussion_r75308039 (patch the Connection)
- [ ] Add missing events from the listener https://github.com/libp2p/js-libp2p-websockets/pull/18/files#r77268889
- [ ] Add documentation about pull-streams
- [ ] Make `main` point to `lib`

**Comment by dignifiedquire** [comment]
> Can't websockets dial timeout? If yes, we need to cb(err)

no timeout by default as far as I know on dial

**Comment by diasdavid** [comment]
interface-transport tests also fail here :( @dignifiedquire

**Comment by dignifiedquire** [comment]
> interface-transport tests also fail here

CI looks pretty green to me after I gave circle a non cache kick

**Pull request merged**